mkosi/firecracker.py

```diff
@@ -0,0 +1,167 @@
+# SPDX-License-Identifier: LGPL-2.1+
+
+import base64
+import contextlib
+import json
+import logging
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+from pathlib import Path
+from typing import Any
+
+from mkosi.config import (
+    Args,
+    Config,
+    ConfigFeature,
+    OutputFormat,
+    QemuFirmware,
+)
+from mkosi.log import die
+from mkosi.partition import finalize_root, find_partitions
+from mkosi.qemu import apply_runtime_size, copy_ephemeral, finalize_drive, generate_scratch_fs, want_scratch
+from mkosi.run import run
+
+
+def maybe_decompress_kernel(src: Path, dst: Path) -> Path:
+    b = src.read_bytes()
```

Comment:
This might be a bit much, since it reads the whole file. It's also a bit weird to do this and then open the file again later. The latter could be avoided by wrapping `b` in a a `io.BytesIO`, but maybe it would be best to open `src` at the beginning?

Reply:
We have to read the whole file regardless to be able to search in it. Opening the file here means an extra level of indentation which I want to avoid.

Reply:
> We have to read the whole file regardless to be able to search in it. 

Yes, but we don't have to read the whole file it into into some buffer we're holding only to read the first four bytes.

> Opening the file here means an extra level of indentation which I want to avoid.

That's true, that's cumbersome, but since we're already opening and closing the file twice, this could just be two with statements on top of each other, adding one level of indentation for the first few lines, but not the whole function.